Automatic TLS Certificates for on-acorn domains

iwilltry42 opened the pull request:
Ref #205 

## To-Do

- [x] Renew if expired
- [x] Set as default for Acorn Managed Ingress
- [x] Get Let's Encrypt details from user input
- [x] Re-Generate if Let's Encrypt settings changed (environment or email)

iwilltry42 commented:
Still hitting some issue that needs fixing:
```
acorn-controller-5d889f4d64-k8xc5 acorn-controller time="2022-09-14T18:50:00Z" level=error msg="error syncing 'acorn-system/acorn-config': handler acorn /v1, Kind=ConfigMap: problem generating wildcard certificate: error: one or more domains had a problem:\n[*.o84oxs.staging.on-acorn.io] [*.o84oxs.staging.on-acorn.io] acme: error presenting token: unexpected response status code: 500\n, requeuing"
```

ibuildthecloud commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,295 @@
+	// Create and Register Let's Encrypt User
+	leUser := &LEUser{
+		Name:  targetUsername,
+		Email: fmt.Sprintf("%s@on-acorn.io", targetUsername),
```
We are going to need to prompt the user during install for an email address. We can't use a generic one, that got us in trouble with Rio when we did that.

ibuildthecloud commented on pkg/publish/ingress.go:
```diff
@@ -19,12 +19,12 @@ import (
 	}
 	if appInstance.Namespace != system.DefaultUserNamespace {
-		hostPrefix += "." + appInstance.Namespace
+		hostPrefix += "---" + appInstance.Namespace
```
What is this for?

iwilltry42 replied:
We can't use dots here since the wildcard cert is only valid one level deep and the triple dash is mostly to avoid ambiguities as mentioned by Vincent here: https://github.com/acorn-io/acorn/issues/205#issuecomment-1240070940 (does that make sense?)

iwilltry42 replied:
Whoops, forgot to add the `<appname>` part. Done.
Example: `http://my-webapp-dawn-moon-9b980585.v4avai.staging.on-acorn.io/`

ibuildthecloud commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,295 @@
+)
+
+const (
+	LEEndpoint = "https://acme-staging-v02.api.letsencrypt.org/directory"
```
This is going to have to come from config.  The only reason the existing DNS was a compile time option is we wanted to change the default value for dev builds. But for LE the default should always be staging and the user should opt into production. We should prompt during install on which one they want (along with email address for production, staging can use a dummy email).

iwilltry42 replied:
Alright, two things here
1. LE staging certs are not trusted by default in browsers.. that's not a problem? Could as well use self-signed certs then, right?
2. Is there some standard way defined already to ask for user input? Also, can one specify a config file upon installation?

iwilltry42 replied:
Oh wow... sorry for my still newborn-son-influenced brain... You probably just mean some new flags, right? :man_facepalming:

ibuildthecloud commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,295 @@
+					logrus.Infof("existing TLS secret %s is still valid until %s", x509crt.Subject.CommonName, x509crt.NotAfter)
+					return sec, nil
+				} else {
+					logrus.Infof("existing TLS secret %s is expired since %s", x509crt.Subject.CommonName, x509crt.NotAfter)
```
Is this how renew works? If so you need to renew the cert before it expires. So something like a week before. Also you can't delete the secret until a new cert is obtained to avoid downtime.

iwilltry42 commented:
![Screenshot-20220919102050-848x645](https://user-images.githubusercontent.com/25345277/190977040-8397e373-1cb7-4fd6-bf2e-da4c284845b0.png)

iwilltry42 commented:
**Default: Let's Encrypt Staging**
![Screenshot-20220922175004-918x774](https://user-images.githubusercontent.com/25345277/191794989-1743f958-a227-4466-b1f7-26efd0f01e0f.png)

**Let's Encrypt Production**

![Screenshot-20220922175333-899x661](https://user-images.githubusercontent.com/25345277/191794978-2f06ec44-a138-4e8b-ac8a-feb65fd10883.png)


![Screenshot-20220922175311-904x383](https://user-images.githubusercontent.com/25345277/191794984-91fae2f2-5ee8-4808-9fff-c191f4d3dc73.png)

iwilltry42 commented on pkg/install/install.go:
```diff
@@ -100,12 +101,36 @@ func DefaultImage() string {
 
+	if opts.Config.LetsEncrypt != nil && *opts.Config.LetsEncrypt == "production" {
+		if opts.Config.LetsEncryptEmail == "" {
+			result, err := pterm.DefaultInteractiveTextInput.WithMultiLine(false).Show("Enter your email address for Let's Encrypt (required for production)")
```
Should this be replaced by something similar to https://github.com/acorn-io/acorn/blob/main/pkg/cli/credential_login.go#L54-L70 ?

cjellick replied:
What you did here looks fine when I run it locally. I don't see a need to to switch to survey

cjellick commented on .github/workflows/main-release.yaml:
```diff
@@ -57,6 +57,7 @@ jobs:
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
           COSIGN_PASSWORD: ${{ secrets.COSIGN_PASSWORD }}
           ACORN_DNS_ENDPOINT: https://staging-dns.acrn.io/v1
+          LE_ENDPOINT: https://acme-staging-v02.api.letsencrypt.org/directory
```
this and the next two files should be unnecessary now, right?

iwilltry42 replied:
Correct, removed.

cjellick commented on pkg/apis/api.acorn.io/v1/types.go:
```diff
@@ -317,6 +317,8 @@ type Config struct {
 	ClusterDomains               []string       `json:"clusterDomains" name:"cluster-domain" usage:"The externally addressable cluster domain (default .on-acorn.io)"`
 	TLSEnabled                   *bool          `json:"tlsEnabled" name:"tls-enabled" usage:"If true HTTPS URLs will be rendered for HTTP endpoint URLs (default false)"`
+	LetsEncrypt                  *string        `json:"letsEncrypt" name:"lets-encrypt" usage:"staging|production|disabled. If set, the Let's Encrypt environment to use for TLS certificates (default staging)"`
+	LetsEncryptEmail             string         `json:"letsEncryptEmail" name:"lets-encrypt-email" usage:"Required if --lets-encrypt=production. The email address to use for Let's Encrypt (default '')"`
```
```suggestion
	LetsEncryptEmail             string         `json:"letsEncryptEmail" name:"lets-encrypt-email" usage:"Required if --lets-encrypt=production. The email address to use for Let's Encrypt registration(default '')"`
```

cjellick commented:
@iwilltry42 - need to do a `make generated` and a `make gen-docs`.

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+package dns
```
Can you go through this file and ensure that only the things that are used outside of package have public visibility?

All the fields on the LEUser struct should be private and many of the functions and methods should also be private.

It'll help know exactly what we're exposing for consumption outside this package.

iwilltry42 replied:
Done.

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+func (u *LEUser) GetEmail() string {
+	return u.Email
+}
+func (u LEUser) GetRegistration() *registration.Resource {
```
This function isn't using a pointer like the others. It should

iwilltry42 replied:
Unused anyway, removed.

iwilltry42 replied:
Changed.

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+	 * Construct new LE User
+	 */
+	targetUsername := strings.TrimPrefix(domain, ".") // leading dot can be an issue
+	email := fmt.Sprintf("%s@on-acorn.io", targetUsername)
```
lets just hard code the staging email to `staging-certs@acorn.io`. I'm going to create that as a real email address so they dont bounce.

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+		return err
+	}
+
+	reg, err := client.Registration.Register(registration.RegisterOptions{TermsOfServiceAgreed: true})
```
We probably need to prompt the user to agree to LE's TOS.

I am thinking that if you say you are doing production LE, you can either pass the flag `--lets-encrypt-tos-agree=true` or if you don't do that, the CLI blocks and prints out a message like:

```
You are choosing to use Let's Encrypt production environment for TLS certificates. To do so, you must agree to their Terms of Serivice: https://community.letsencrypt.org/tos

Do you agree to Let's Encrypt TOS? y/n
```

Thoughts?

iwilltry42 replied:
Makes total sense, done!

cjellick commented on pkg/system/constants.go:
```diff
@@ -5,6 +5,8 @@ import "os"
 	Namespace            = "acorn-system"
 	ConfigName           = "acorn-config"
+	TLSSecretName        = "acorn-tls"
+	LESecretName         = "acorn-le"
```
For clarity, can you add "account" to this?
```suggestion
	LEAccountSecretName         = "acorn-le-account"
```

cjellick commented on pkg/dns/daemon.go:
```diff
@@ -82,6 +82,8 @@ func (d *Daemon) internal(ctx context.Context) (bool, error) {
 		return false, nil
 	}
 
+	// TODO: renew certificate
```
Yeah, I think renewal should happen here, not in the handler. Any reason you can't move that logic here? Would be fine if its in both places

cjellick replied:
changed my mind. not necessary. the controller is good enough for now i think

cjellick commented on pkg/controller/dns/config.go:
```diff
@@ -63,6 +63,20 @@ func (h *configHandler) Handle(req router.Request, resp router.Response) error {
+		}
+
+		// Generate wildcard certificate for domain
+		_, err = leUser.EnsureWildcardCertificateSecret(req.Ctx, req.Client, *cfg.AcornDNSEndpoint, domain, token)
```
said elsewhere, but i think we need to also handle renew in the dns daemon, not just the handler

cjellick replied:
changed my mind. this is good enough. dont need to also do in a daemon

cjellick commented on pkg/controller/dns/provider.go:
```diff
@@ -0,0 +1,86 @@
+	client DNSClient
+}
+
+func NewDNSProvider(endpoint, domain, token string) *DNSProvider {
```
For clarity, can you rename this `NewACMEDNS01ChallengeProvider`. It wasn't clear to me what it was for at first

iwilltry42 replied:
Done.

cjellick commented on pkg/dns/helpers.go:
```diff
@@ -9,7 +9,7 @@ import (
 	"github.com/sirupsen/logrus"
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/api/networking/v1"
+	v1 "k8s.io/api/networking/v1"
```
weird change. can you eliminate it?

iwilltry42 replied:
For some reason, vscode adds this automatically on save. Removed it with vim for now :D

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      system.TLSSecretName,
+			Namespace: system.Namespace,
+			Annotations: map[string]string{
```
Of these annotations, only the hash is ever read. What is the intention for the other three? Just informational?

iwilltry42 replied:
They make debugging easier in case anything goes wrong, but I'm totally fine with removing them as they're only informational.

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,349 @@
+
+	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)
+
+	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
```
As mentioned in slack, I ran into a lot of trouble with my local DNS resolver trying to do the "pre check" for DNS. Not sure if you caught that bit of the convo, but this client attempts to ensure that the TXT record is there locally before telling let's encrypt to do it. The problem is that local dns resolvers can act really funky and for me, they never succeed. We agreed that if we just wait 60 seconds (the documented time for dns to proprogate throughout route53), we can just skip the local check entirely.

Here is the diff that allowed me to achieve this locally:
```
diff --git a/pkg/controller/dns/letsencrypt.go b/pkg/controller/dns/letsencrypt.go
index c5af25f..a17b690 100644
--- a/pkg/controller/dns/letsencrypt.go
+++ b/pkg/controller/dns/letsencrypt.go
@@ -21,6 +21,7 @@ import (
 	"github.com/acorn-io/baaah/pkg/router"
 	"github.com/go-acme/lego/v4/certcrypto"
 	"github.com/go-acme/lego/v4/certificate"
+	"github.com/go-acme/lego/v4/challenge/dns01"
 	"github.com/go-acme/lego/v4/lego"
 	"github.com/go-acme/lego/v4/registration"
 	"github.com/sirupsen/logrus"
@@ -143,6 +144,10 @@ func (u *LEUser) ToSecret() (*corev1.Secret, error) {
 	}, nil
 }

+func noOpCheck(_, _, _ string, _ dns01.PreCheckFunc) (bool, error) {
+	return true, nil
+}
+
 func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certificate.Resource, error) {
 	if u.Registration == nil {
 		return nil, fmt.Errorf("not generating LE cert: missing registration")
@@ -163,7 +168,7 @@ func (u *LEUser) GenerateWildcardCert(dnsendpoint, domain, token string) (*certi

 	dnsProvider := NewDNSProvider(dnsendpoint, domain, token)

-	if err := client.Challenge.SetDNS01Provider(dnsProvider); err != nil {
+	if err := client.Challenge.SetDNS01Provider(dnsProvider, dns01.WrapPreCheck(noOpCheck)); err != nil {
 		return nil, err
 	}

diff --git a/pkg/controller/dns/provider.go b/pkg/controller/dns/provider.go
index 93c1b20..612167d 100644
--- a/pkg/controller/dns/provider.go
+++ b/pkg/controller/dns/provider.go
@@ -40,7 +40,7 @@ func (d *DNSProvider) CleanUp(domain, token, keyAuth string) error {
 }

 func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
-	return 30 * time.Second, 5 * time.Second
+	return 3 * time.Minute, 1 * time.Minut
 }

```

meldafrawi commented:
if user install acorn **without** `--tls-enabled --lets-encrypt --lets-encrypt-email`  and wanted to enable LE later, should the user just reinstall with the flags? will it require any restart on acorn api or controller ?

meldafrawi commented on pkg/install/install.go:
```diff
@@ -100,12 +101,36 @@ func DefaultImage() string {
+			if err != nil {
+				return err
+			}
+			opts.Config.LetsEncryptEmail = result
```
@iwilltry42 
if email is required when using Let's Encrypt  `production`, should we check email address is valid before setting `opts.Config.LetsEncryptEmail`, and abort installation if not valid ? 

cc: @cjellick

iwilltry42 replied:
We're checking it outside of this `if` statement so it's checking the validity for production **and** staging.
It will error out in both scenarios if the email is invalid: https://github.com/acorn-io/acorn/pull/693/files#diff-3c83c1219d23236aa4c92d672faa69bb463e2001adb87833ca0f7434faded420R127-R131

iwilltry42 replied:
Moved it even further down to always check it if it was set, either via flag (for both prod/staging) or via interactive input (prompt only for prod)

cjellick commented:
Ok, Here is the summary of todos/changes/comments, adapted from what you sent me
- Update host prefix handling according to https://github.com/acorn-io/acorn/pull/693#discussion_r981859831
- When configured to use staging LE, set the email to staging-certs@acorn.io. If the user supplies an email anyway, respect it https://github.com/acorn-io/acorn/pull/693#discussion_r981815386
- Cert loop should not block/error-out DNS config handler. This is currently causing it to spam acorn-dns with new domain requests
- ~Cert Refresh will be handled in a Daemon as well~ Per our slack convo and my comments here, I changed my mind. The current approach is good enough, like you said
- Other PR comments, like changing the visibility of functions/variables in the letsencrypt.go file

If you wanted to get the above core functionality done and get the PR merged and then follow up on the below in a separate PR, I'd be fine with that, up to you.
- If the user opts-in to product let's encrypt, do the following:
  - Prompt the user to agree to the TOS per:  https://github.com/acorn-io/acorn/pull/693#pullrequestreview-1122851564
  - Add a message in the install output saying "You've enabled automatic TLS certificates with Let's Encrypt. This can take a few minutes to configure"
  - Add something to acorn info to say whether or not the certificate has been issued. Can be as simple as `letsEncrypt: production, pending`, `letsEncrypt: production, provisioned`
  - The URLs in `acorn ps` should start with https. Doing this to all endpoints actually is probably an oversimplification. So, if you can figure out a more elegant solution to do it on a per endpoint basis, great. But I don't want that to hold up shipping this feature this week. All that logic is here: https://github.com/acorn-io/acorn/blob/main/pkg/controller/appdefinition/status.go#L390-L397

iwilltry42 commented:
> if user install acorn **without** `--tls-enabled --lets-encrypt --lets-encrypt-email` and wanted to enable LE later, should the user just reinstall with the flags? will it require any restart on acorn api or controller ?

I think this was clarified via Slack already, but no, it won't require restarts as it's hooked up in the config handler and secret changes are propagated transparently. :magic_wand:

cjellick commented on pkg/apis/api.acorn.io/v1/types.go:
```diff
@@ -316,7 +317,9 @@ type Config struct {
 	ClusterDomains               []string       `json:"clusterDomains" name:"cluster-domain" usage:"The externally addressable cluster domain (default .on-acorn.io)"`
-	TLSEnabled                   *bool          `json:"tlsEnabled" name:"tls-enabled" usage:"If true HTTPS URLs will be rendered for HTTP endpoint URLs (default false)"`
+	LetsEncrypt                  *string        `json:"letsEncrypt" name:"lets-encrypt" usage:"enabled|disabled|staging. If enabled, uses the Let's Encrypt production environment to issue a wildcard TLS certificate. (default disabled)"`
+	LetsEncryptEmail             string         `json:"letsEncryptEmail" name:"lets-encrypt-email" usage:"Required if --lets-encrypt=production. The email address to use for Let's Encrypt registration(default '')"`
```
```suggestion
	LetsEncryptEmail             string         `json:"letsEncryptEmail" name:"lets-encrypt-email" usage:"Required if --lets-encrypt=enabled. The email address to use for Let's Encrypt registration(default '')"`
```

cjellick commented on pkg/apis/api.acorn.io/v1/types.go:
```diff
@@ -316,7 +317,9 @@ type Config struct {
-	TLSEnabled                   *bool          `json:"tlsEnabled" name:"tls-enabled" usage:"If true HTTPS URLs will be rendered for HTTP endpoint URLs (default false)"`
+	LetsEncrypt                  *string        `json:"letsEncrypt" name:"lets-encrypt" usage:"enabled|disabled|staging. If enabled, uses the Let's Encrypt production environment to issue a wildcard TLS certificate. (default disabled)"`
+	LetsEncryptEmail             string         `json:"letsEncryptEmail" name:"lets-encrypt-email" usage:"Required if --lets-encrypt=production. The email address to use for Let's Encrypt registration(default '')"`
+	LetsEncryptTOSAgree          *bool          `json:"letsEncryptTOSAgree" name:"lets-encrypt-tos-agree" usage:"Required if --lets-encrypt=production. If true, you agree to the Let's Encrypt terms of service (default false)"`
```
```suggestion
	LetsEncryptTOSAgree          *bool          `json:"letsEncryptTOSAgree" name:"lets-encrypt-tos-agree" usage:"Required if --lets-encrypt=enabled. If true, you agree to the Let's Encrypt terms of service (default false)"`
```

cjellick commented on pkg/config/config.go:
```diff
@@ -50,6 +51,20 @@ func complete(c *apiv1.Config, ctx context.Context, getter kclient.Reader) error
+	}
+	if *c.LetsEncrypt == "enabled" {
+		if c.LetsEncryptEmail == "" {
+			return fmt.Errorf("letsencrypt email is required when using production")
```
```suggestion
			return fmt.Errorf("letsencrypt email is required when Let's Encrypt is enabled")
```

cjellick commented on pkg/config/config.go:
```diff
@@ -50,6 +51,20 @@ func complete(c *apiv1.Config, ctx context.Context, getter kclient.Reader) error
+			return fmt.Errorf("letsencrypt email is required when using production")
+		}
+		if !*c.LetsEncryptTOSAgree {
+			return fmt.Errorf("letsencrypt TOS must be agreed to when using production")
```
```suggestion
			return fmt.Errorf("letsencrypt TOS must be agreed to when Let's Encrypt is enabled")
```

cjellick commented on pkg/controller/dns/letsencrypt.go:
```diff
@@ -0,0 +1,354 @@
+	if strings.EqualFold(*cfg.LetsEncrypt, "enabled") {
+		url = LetsEncryptURLProduction
+		if cfg.LetsEncryptEmail == "" {
+			return nil, fmt.Errorf("let's encrypt email is required for production")
```
```suggestion
			return nil, fmt.Errorf("let's encrypt email is required")
```

cjellick commented on pkg/install/install.go:
```diff
@@ -100,12 +101,51 @@ func DefaultImage() string {
+			opts.Config.LetsEncryptTOSAgree = &ok
+		}
+		if opts.Config.LetsEncryptEmail == "" {
+			result, err := pterm.DefaultInteractiveTextInput.WithMultiLine(false).Show("Enter your email address for Let's Encrypt (required for production)")
```
```suggestion
			result, err := pterm.DefaultInteractiveTextInput.WithMultiLine(false).Show("Enter your email address for Let's Encrypt")
```

cjellick commented on pkg/install/install.go:
```diff
@@ -100,12 +101,51 @@ func DefaultImage() string {
+			}
+			opts.Config.LetsEncryptEmail = result
+		}
+		pterm.Info.Println("You've enabled automatic TLS certificate provisioning with Let's Encrypt production. This can take a few minutes to configure.")
```
```suggestion
		pterm.Info.Println("You've enabled automatic TLS certificate provisioning with Let's Encrypt. This can take a few minutes to configure.")
```

cjellick commented on pkg/controller/appdefinition/status.go:
```diff
@@ -369,13 +375,34 @@ func podName(pod *corev1.Pod) string {
+		return "", err
+	}
+
+	ingressTLSHosts := []string{}
```
nitpick, a map would be more efficent. If this where a `map[string]interface{}`, line 423 would be a much more efficient lookup.

iwilltry42 replied:
Yup :)

cjellick commented on pkg/install/install.go:
```diff
@@ -100,12 +101,51 @@ func DefaultImage() string {
+	}
+
+	// Validate E-Mail address provided for Let's Encrypt registration
+	if opts.Config.LetsEncryptEmail != "" {
```
Found a bug with this: If I use the prompts to enter my email, it lets me just hit enter to submit a blank email.
That lets the install continue with a blank email and then the controller gets in an error loop.

So, this check probably needs to add an "or" for when letsEncrypt is enabled, like `if opts.Config.LetsEncryptEmail != "" || opts.COnfig.LetsEncrypt == "enabled {`

iwilltry42 replied:
Good catch! Alternatively we could make `LetsEncryptEmail` a `*string` type flag as well and check for `nil` instead of `""`.
Went for or check for now :heavy_check_mark:

cjellick commented on pkg/install/install.go:
```diff
@@ -100,12 +101,51 @@ func DefaultImage() string {
+	// Require E-Mail address when using Let's Encrypt production
+	if opts.Config.LetsEncrypt != nil && *opts.Config.LetsEncrypt == "enabled" {
+		if opts.Config.LetsEncryptTOSAgree == nil || !*opts.Config.LetsEncryptTOSAgree {
+			ok, err := pterm.DefaultInteractiveConfirm.Show("You are choosing to use Let's Encrypt production environment for TLS certificates. To do so, you must agree to their Terms of Service: https://community.letsencrypt.org/tos\nTip: use --lets-encrypt-tos-agree to skip this prompt\nDo you agree to Let's Encrypt TOS?")
```
There is another thing I don't like about this: as soon as you enter 'y', it continues. It doesnt wait for you to hit enter. I think it should. Our other yes/no prompts (most obvious when you do `acorn uninstall -a`) wait until you hit enter.

This caused me to stumble because I hit y and then enter, but enter actually applied to the next prompt, which was for my email and it caused be to submit a blank email

iwilltry42 commented:
Rebased, fixed conflicts and re-generated docs.
`make generate` still does not generate the openapi stuff for me, so maybe one of you can look into this?
